### ISSUE TITLE
(omake crash) How compatible omake is to gnu make?

### ISSUE THREAD

**Issue by ghost** [LADSoft/OrangeC#792]
This the the makefile generated by the DevCpp IDE for `mingw32-make`:

```
# Project: DuiLib
# Makefile created by Red Panda C++ 2.10

CPP      = clang++.exe
CC       = clang.exe
WINDRES  = windres.exe
OBJ      = Utils/WinImplBase.o Control/UIScrollBar.o Control/UILabel.o Utils/UIDelegate.o Utils/stb_image.o Control/UIDateTime.o Control/UIProgress.o Control/UIFlash.o Control/UIEdit.o Layout/UITabLayout.o Layout/UIChildLayout.o Core/UIMarkup.o Core/UIManager.o Control/UIOption.o StdAfx.o Core/UIControl.o UIlib.o Control/UISlider.o Utils/WndShadow.o Layout/UIVerticalLayout.o Layout/UITileLayout.o Control/UIWebBrowser.o Control/UIActiveX.o Control/UIText.o Core/UIRender.o Control/UIGifAnim.o Control/UICombo.o Core/UIDlgBuilder.o Control/UIButton.o Layout/UIHorizontalLayout.o Control/UITreeView.o Utils/Utils.o Control/UIRichEdit.o Control/UICheckBox.o Control/UIList.o Utils/XUnzip.o Core/UIContainer.o Core/UIBase.o
LINKOBJ  = Utils/WinImplBase.o Control/UIScrollBar.o Control/UILabel.o Utils/UIDelegate.o Utils/stb_image.o Control/UIDateTime.o Control/UIProgress.o Control/UIFlash.o Control/UIEdit.o Layout/UITabLayout.o Layout/UIChildLayout.o Core/UIMarkup.o Core/UIManager.o Control/UIOption.o StdAfx.o Core/UIControl.o UIlib.o Control/UISlider.o Utils/WndShadow.o Layout/UIVerticalLayout.o Layout/UITileLayout.o Control/UIWebBrowser.o Control/UIActiveX.o Control/UIText.o Core/UIRender.o Control/UIGifAnim.o Control/UICombo.o Core/UIDlgBuilder.o Control/UIButton.o Layout/UIHorizontalLayout.o Control/UITreeView.o Utils/Utils.o Control/UIRichEdit.o Control/UICheckBox.o Control/UIList.o Utils/XUnzip.o Core/UIContainer.o Core/UIBase.o
CLEANOBJ  =  Utils\WinImplBase.o Control\UIScrollBar.o Control\UILabel.o Utils\UIDelegate.o Utils\stb_image.o Control\UIDateTime.o Control\UIProgress.o Control\UIFlash.o Control\UIEdit.o Layout\UITabLayout.o Layout\UIChildLayout.o Core\UIMarkup.o Core\UIManager.o Control\UIOption.o StdAfx.o Core\UIControl.o UIlib.o Control\UISlider.o Utils\WndShadow.o Layout\UIVerticalLayout.o Layout\UITileLayout.o Control\UIWebBrowser.o Control\UIActiveX.o Control\UIText.o Core\UIRender.o Control\UIGifAnim.o Control\UICombo.o Core\UIDlgBuilder.o Control\UIButton.o Layout\UIHorizontalLayout.o Control\UITreeView.o Utils\Utils.o Control\UIRichEdit.o Control\UICheckBox.o Control\UIList.o Utils\XUnzip.o Core\UIContainer.o Core\UIBase.o libDuiLib.a
LIBS     = 
INCS     =  
CXXINCS  =  
BIN      = libDuiLib.a
CXXFLAGS = $(CXXINCS)  -g3 -w -Og -std=c++98 -pipe -DUNICODE -D_UNICODE -I. -fpermissive -D__DEBUG__
CFLAGS   = $(INCS)  -g3 -w -Og -pipe -D__DEBUG__
RM       = del /q /f
WINDRESFLAGS  = 

.PHONY: all all-before all-after clean clean-custom

all: all-before $(BIN) all-after

clean: clean-custom
	-$(RM) $(CLEANOBJ) > NUL 2>&1

$(BIN): $(LINKOBJ)
	ar r $(BIN) $(LINKOBJ)
	ranlib $(BIN)

Utils/WinImplBase.o: Utils/WinImplBase.cpp
	$(CPP) -c Utils/WinImplBase.cpp -o Utils/WinImplBase.o $(CXXFLAGS) 

Control/UIScrollBar.o: Control/UIScrollBar.cpp Control/UIScrollBar.h
	$(CPP) -c Control/UIScrollBar.cpp -o Control/UIScrollBar.o $(CXXFLAGS) 

Control/UILabel.o: Control/UILabel.cpp Control/UILabel.h
	$(CPP) -c Control/UILabel.cpp -o Control/UILabel.o $(CXXFLAGS) 

Utils/UIDelegate.o: Utils/UIDelegate.cpp
	$(CPP) -c Utils/UIDelegate.cpp -o Utils/UIDelegate.o $(CXXFLAGS) 

Utils/stb_image.o: Utils/stb_image.c Utils/stb_image.h
	$(CC) -c Utils/stb_image.c -o Utils/stb_image.o $(CFLAGS) 

Control/UIDateTime.o: Control/UIDateTime.cpp Control/UIDateTime.h
	$(CPP) -c Control/UIDateTime.cpp -o Control/UIDateTime.o $(CXXFLAGS) 

Control/UIProgress.o: Control/UIProgress.cpp Control/UIProgress.h
	$(CPP) -c Control/UIProgress.cpp -o Control/UIProgress.o $(CXXFLAGS) 

Control/UIFlash.o: Control/UIFlash.cpp Control/UIFlash.h
	$(CPP) -c Control/UIFlash.cpp -o Control/UIFlash.o $(CXXFLAGS) 

Control/UIEdit.o: Control/UIEdit.cpp Control/UIEdit.h
	$(CPP) -c Control/UIEdit.cpp -o Control/UIEdit.o $(CXXFLAGS) 

Layout/UITabLayout.o: Layout/UITabLayout.cpp Layout/UITabLayout.h
	$(CPP) -c Layout/UITabLayout.cpp -o Layout/UITabLayout.o $(CXXFLAGS) 

Layout/UIChildLayout.o: Layout/UIChildLayout.cpp Layout/UIChildLayout.h
	$(CPP) -c Layout/UIChildLayout.cpp -o Layout/UIChildLayout.o $(CXXFLAGS) 

Core/UIMarkup.o: Core/UIMarkup.cpp
	$(CPP) -c Core/UIMarkup.cpp -o Core/UIMarkup.o $(CXXFLAGS) 

Core/UIManager.o: Core/UIManager.cpp
	$(CPP) -c Core/UIManager.cpp -o Core/UIManager.o $(CXXFLAGS) 

Control/UIOption.o: Control/UIOption.cpp Control/UIOption.h
	$(CPP) -c Control/UIOption.cpp -o Control/UIOption.o $(CXXFLAGS) 

StdAfx.o: StdAfx.cpp Control/UIRichEdit.h Layout/UIChildLayout.h UIlib.h Control/UICombo.h Control/UIButton.h Core/UIRender.h Core/UIMarkup.h Control/UILabel.h Control/UIGifAnim.h Core/UIControl.h Core/UIContainer.h Control/UIProgress.h Control/UIScrollBar.h Control/UIOption.h Core/UIBase.h Control/UIActiveX.h Control/UIEdit.h Core/UIDefine.h Control/UISlider.h Utils/WinImplBase.h Control/UIText.h Control/UIList.h Core/UIManager.h Core/UIDlgBuilder.h Control/UIWebBrowser.h Control/UIDateTime.h StdAfx.h Layout/UIHorizontalLayout.h Utils/WndShadow.h Control/UITreeView.h Layout/UIVerticalLayout.h Utils/Utils.h Layout/UITileLayout.h Control/UICheckBox.h Layout/UITabLayout.h Utils/UIDelegate.h
	$(CPP) -c StdAfx.cpp -o StdAfx.o $(CXXFLAGS) 

Core/UIControl.o: Core/UIControl.cpp
	$(CPP) -c Core/UIControl.cpp -o Core/UIControl.o $(CXXFLAGS) 

UIlib.o: UIlib.cpp Control/UIRichEdit.h Layout/UIChildLayout.h UIlib.h Control/UICombo.h Control/UIButton.h Core/UIRender.h Core/UIMarkup.h Control/UILabel.h Control/UIGifAnim.h Core/UIControl.h Core/UIContainer.h Control/UIProgress.h Control/UIScrollBar.h Control/UIOption.h Core/UIBase.h Control/UIActiveX.h Control/UIEdit.h Core/UIDefine.h Control/UISlider.h Utils/WinImplBase.h Control/UIText.h Control/UIList.h Core/UIManager.h Core/UIDlgBuilder.h Control/UIWebBrowser.h Control/UIDateTime.h Layout/UIHorizontalLayout.h Utils/WndShadow.h Control/UITreeView.h Layout/UIVerticalLayout.h Utils/Utils.h Layout/UITileLayout.h Control/UICheckBox.h Layout/UITabLayout.h Utils/UIDelegate.h
	$(CPP) -c UIlib.cpp -o UIlib.o $(CXXFLAGS) 

Control/UISlider.o: Control/UISlider.cpp Control/UISlider.h
	$(CPP) -c Control/UISlider.cpp -o Control/UISlider.o $(CXXFLAGS) 

Utils/WndShadow.o: Utils/WndShadow.cpp Utils/WndShadow.h
	$(CPP) -c Utils/WndShadow.cpp -o Utils/WndShadow.o $(CXXFLAGS) 

Layout/UIVerticalLayout.o: Layout/UIVerticalLayout.cpp Layout/UIVerticalLayout.h
	$(CPP) -c Layout/UIVerticalLayout.cpp -o Layout/UIVerticalLayout.o $(CXXFLAGS) 

Layout/UITileLayout.o: Layout/UITileLayout.cpp Layout/UITileLayout.h
	$(CPP) -c Layout/UITileLayout.cpp -o Layout/UITileLayout.o $(CXXFLAGS) 

Control/UIWebBrowser.o: Control/UIWebBrowser.cpp Control/UIWebBrowser.h Utils/downloadmgr.h
	$(CPP) -c Control/UIWebBrowser.cpp -o Control/UIWebBrowser.o $(CXXFLAGS) 

Control/UIActiveX.o: Control/UIActiveX.cpp
	$(CPP) -c Control/UIActiveX.cpp -o Control/UIActiveX.o $(CXXFLAGS) 

Control/UIText.o: Control/UIText.cpp Control/UIText.h
	$(CPP) -c Control/UIText.cpp -o Control/UIText.o $(CXXFLAGS) 

Core/UIRender.o: Core/UIRender.cpp
	$(CPP) -c Core/UIRender.cpp -o Core/UIRender.o $(CXXFLAGS) 

Control/UIGifAnim.o: Control/UIGifAnim.cpp Control/UIGifAnim.h
	$(CPP) -c Control/UIGifAnim.cpp -o Control/UIGifAnim.o $(CXXFLAGS) 

Control/UICombo.o: Control/UICombo.cpp
	$(CPP) -c Control/UICombo.cpp -o Control/UICombo.o $(CXXFLAGS) 

Core/UIDlgBuilder.o: Core/UIDlgBuilder.cpp
	$(CPP) -c Core/UIDlgBuilder.cpp -o Core/UIDlgBuilder.o $(CXXFLAGS) 

Control/UIButton.o: Control/UIButton.cpp Control/UIButton.h
	$(CPP) -c Control/UIButton.cpp -o Control/UIButton.o $(CXXFLAGS) 

Layout/UIHorizontalLayout.o: Layout/UIHorizontalLayout.cpp Layout/UIHorizontalLayout.h
	$(CPP) -c Layout/UIHorizontalLayout.cpp -o Layout/UIHorizontalLayout.o $(CXXFLAGS) 

Control/UITreeView.o: Control/UITreeView.cpp Control/UITreeView.h
	$(CPP) -c Control/UITreeView.cpp -o Control/UITreeView.o $(CXXFLAGS) 

Utils/Utils.o: Utils/Utils.cpp Utils/Utils.h
	$(CPP) -c Utils/Utils.cpp -o Utils/Utils.o $(CXXFLAGS) 

Control/UIRichEdit.o: Control/UIRichEdit.cpp
	$(CPP) -c Control/UIRichEdit.cpp -o Control/UIRichEdit.o $(CXXFLAGS) 

Control/UICheckBox.o: Control/UICheckBox.cpp Control/UICheckBox.h
	$(CPP) -c Control/UICheckBox.cpp -o Control/UICheckBox.o $(CXXFLAGS) 

Control/UIList.o: Control/UIList.cpp
	$(CPP) -c Control/UIList.cpp -o Control/UIList.o $(CXXFLAGS) 

Utils/XUnzip.o: Utils/XUnzip.cpp
	$(CPP) -c Utils/XUnzip.cpp -o Utils/XUnzip.o $(CXXFLAGS) 

Core/UIContainer.o: Core/UIContainer.cpp
	$(CPP) -c Core/UIContainer.cpp -o Core/UIContainer.o $(CXXFLAGS) 

Core/UIBase.o: Core/UIBase.cpp
	$(CPP) -c Core/UIBase.cpp -o Core/UIBase.o $(CXXFLAGS) 

```

The project it supposed to build is: https://github.com/duilib/duilib/tree/master/DuiLib

The problem: omake will complain about `all-after` and `all-before`, after removed them omake crashed without giving a reasons.

**Comment by GitMensch** [comment]
The global question can be answered globally by "Omake is quite compatible to GNU make".

but there's an actual issue in here:

> after removed them omake crashed without giving a reasons.

Whoa. Can you please check with the latest debug build if you get more than a "crash without reason"?
Current latest is https://ci.appveyor.com/api/buildjobs/sn1my6p725veuqb7/artifacts/dist%2FZippedBinaries60626.7z

as this can be tracked here, please consider renaming this issue to "omake chrash" or similar.

Separate to this...
> The problem: omake will complain about all-after and all-before

this seems related to #790 - broken support for .PHONY target.

**Comment by tn1997tn** [comment]
> this seems related to #790 - broken support for .PHONY target.

I don't think it's that simple. You are welcome to debug it further, I gave up.

**Comment by LADSoft** [comment]
will look into this tonite or tomorrow

**Comment by LADSoft** [comment]
@tn1997tn

I still cant get omake to crash.    What os are you using?   What os version?   What command shell?

BTW when I set the compiler to C++14 in the make file the microsoft compilers would not compile this project because of extra qualifications in qualified names...   setting it to C++11 doesn't work either because the project does use C++14....

**Comment by tn1997tn** [comment]
> @tn1997tn
> 
> I still cant get omake to crash. What os are you using? What os version? What command shell?

Windows 8.1 64 bit. Windows Command Prompt.

> BTW when I set the compiler to C++14 in the make file the microsoft compilers would not compile this project because of extra qualifications in qualified names... setting it to C++11 doesn't work either because the project does use C++14....

That project is very buggy, don't expect it to compile. The problem we are discussing here is omake, it crashed.

**Comment by LADSoft** [comment]
ok that helps, I've been testing solely on windows 10/11 lately...   I will see if I can find my windows 8.1 box and try it out...

**Comment by LADSoft** [comment]
my windows 8 box is inaccessible because I didn't have the password...   i did try it on windows 7 and it seems to work though.    Meanwhile I analyzed the code for omake and found a couple of bugs, using -j 0 crashed it outright and without a -j switch it did some questionable things rather than default correctly to -j:1...   I suppose that those things could have have crashed it under certain circumstances but don't really know...    I've fixed those issues...   if you feel inclined to try it I would be interested in hearing any results...

**Comment by chuggafan** [comment]
> -j switch it did some questionable things rather than default correctly to -j:1...

I swear we had a discussion about this behavior and this behavior is intentional so to allow an infinite number of threads to be dispatched across a build cluster rather than one single machine.

**Comment by LADSoft** [comment]
@chuggafan

according to the GNU documentation, -j without a number leads to that behavior; you get infinite threads.    Also according to GNU documentation, not having a -j switch at all should be equivalent to -j:1.   It is the latter behavior which was broken.    FWIW I think from reading the code it was your intent to handle the -j switch properly, however, I think I broke it when I did that rewrite last summer or fall...
